File: python/completed/two_sum.py
'''
Problem 1

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
'''

class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        summation = 0
        sum_not_found = True
        starting_index = 0


        while sum_not_found:
            for i in range(starting_index+1, len(nums)):
                summation = nums[starting_index] + nums[i]
                if summation == target:
                    sum_not_found = False
                    break

            if sum_not_found is False:
                pass
            else:
                starting_index += 1

        if sum_not_found is False:
            return [starting_index, i]

    def twoSum2(self, nums, target):
        # key: target-number
        # value: number index
        sum_dict = {}

        for i in range(len(nums)):
            if nums[i] in sum_dict.keys():
                return [sum_dict[nums[i]], i]
            else:
                sum_dict[target-nums[i]] = i

if __name__ == '__main__':
    numbers = [3,3]
    t = 6

    s = Solution()
    r = s.twoSum(numbers, t)
    print(r)

File: js/two_sum.js
/**
 * Problem 1
 * 
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    var summation = 0;
    var starting_index = 0;
    var sum_not_found = true;

    while (sum_not_found) {
        for (var i = starting_index + 1; i < nums.length; i++) {
            var summation = nums[starting_index] + nums[i];
            console.log(starting_index, i);
            if (summation === target){
                var sum_not_found = false;
                console.log(sum_not_found);
                break
            }
        }
        if (sum_not_found) {
                starting_index++;
        }
    }
    if (!sum_not_found){
        var indices = [starting_index, i];
        return indices;
    }
};

var twoSum2 = function(nums, target) {
    var sum_map = new Map();

    for (let i = 0; i < nums.length; i++) {
        if (sum_map.has(nums[i])) {
            let result = [sum_map.get(nums[i]), i];
            return result;
        } else {
            sum_map.set(target-nums[i], i);
        }
    }
}

let nums = [3, 2, 3];
let target = 6;
let result = twoSum(nums, target);
console.log(result);